src/object.c

```diff
@@ -190,18 +190,21 @@ dispatch_release(dispatch_object_t dou)
 void
 _dispatch_xref_dispose(dispatch_object_t dou)
 {
-	unsigned long metatype = dx_metatype(dou._do);
-	if (metatype == _DISPATCH_QUEUE_TYPE || metatype == _DISPATCH_SOURCE_TYPE) {
+	if (dx_cluster(dou._do) == _DISPATCH_QUEUE_CLUSTER) {
 		_dispatch_queue_xref_dispose(dou._dq);
 	}
-	if (dx_type(dou._do) == DISPATCH_SOURCE_KEVENT_TYPE) {
+	switch (dx_type(dou._do)) {
+	case DISPATCH_SOURCE_KEVENT_TYPE:
 		_dispatch_source_xref_dispose(dou._ds);
+		break;
 #if HAVE_MACH
-	} else if (dx_type(dou._do) == DISPATCH_MACH_CHANNEL_TYPE) {
+	case DISPATCH_MACH_CHANNEL_TYPE:
 		_dispatch_mach_xref_dispose(dou._dm);
+		break;
 #endif
-	} else if (dx_type(dou._do) == DISPATCH_QUEUE_RUNLOOP_TYPE) {
-		_dispatch_runloop_queue_xref_dispose(dou._dq);
+	case DISPATCH_QUEUE_RUNLOOP_TYPE:
+		_dispatch_runloop_queue_xref_dispose(dou._dl);
```

Comment:
Not sure why this wasn't a problem before, but `_dispatch_runloop_queue_xref_dispose` is conditional to `DISPATCH_COCOA_COMPAT`.

Reply:
I think that branch of the switch should also be under DISPATCH_COCOA_COMPAT.